### PR TITLE
DiscoveryService: prevent EICE Node duplication

### PR DESCRIPTION
**Pull request by marcoandredinis**
This PR ensures does a couple of things to prevent duplicated nodes.

### Run only one DiscoveryService per DiscoveryGroup
This should prevent any duplicate API calls and also prevent duplicate resources on Teleport side.

### Deterministic name for EICE Nodes
Instead of using a random UUID, we are now building the Node's Name based on the account id and instance ID of the EC2 instance.

This reduces the load because we can now use the NodeWatcher's internal Map for quicker access (instead of always listing and filtering all the Nodes).

Given the name is deterministic, it also ensures we don't get duplicate nodes.

### CODE REVIEW

File: api/types/server.go
Comment (espadolini) [inline api/types/server.go:510]: Does the instance ID always include the `i-` part?

File: api/types/server.go
Comment (espadolini) [inline api/types/server.go:513]: Should this be an error or an empty string checked by the caller, instead?

File: lib/srv/discovery/discovery.go
Comment (espadolini) [inline lib/srv/discovery/discovery.go:1327]: Do we need to bother with also canceling all in-progress operations with a context, or is this mostly to avoid doing the same work twice?

File: lib/srv/discovery/discovery.go
Comment (espadolini) [inline lib/srv/discovery/discovery.go:1381]: Do we need to synchronize the release function, from a memory model perspective and from a logical one? What if we end up stopping the `*Server` before we managed to update the `releaseGroupLockFn`?

File: api/types/server.go
Comment (marcoandredinis) [reply api/types/server.go:510]: It should, yes.
We don't validate it but AWS is always returning the `i-` part as far as I can tell.

File: api/types/server.go
Comment (marcoandredinis) [reply api/types/server.go:513]: This is only called in the CheckAndSetDefaults, which ends up checking if the Server has the Account and Instance ID.
So, even if we default to an uuid, it will fail afterwards because one (or both) of those values are missing.

I added the check anyway so that we can fail earlier.

File: lib/srv/discovery/discovery.go
Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go:1381]: I moved the `releaseGroupLockFn` creation to much earlier.
Now, it's the first thing it does when calling the `Start`

This should ensure it's very unlikely to call the `Stop` before `releaseGroupLockFn` was set.

File: api/types/server.go
Comment (espadolini) [inline api/types/server.go:528]: Should we go heavy-duty here and use `api/utils/aws.IsEC2NodeID`?

File: lib/srv/discovery/discovery.go
Comment (espadolini) [inline lib/srv/discovery/discovery.go:1346]: From what I can tell there's a potential race condition on `s.releaseGroupLockFn` and a potential race condition on `lease`.

The logical race condition of `(*Server).Stop` getting called right before or during a call to `acquireDiscoveryGroup`, leading to resources that are never going to get cleaned up is also still there I think?

File: api/types/server.go
Comment (marcoandredinis) [reply api/types/server.go:528]: I'll add a test and a comment here

File: lib/srv/discovery/discovery.go
Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go]: Can you please take another look?
I think I fixed both races, and `go test -race ...` seems to agree.

File: lib/srv/discovery/discovery.go
Comment (tigrato) [inline lib/srv/discovery/discovery.go:1412]:
```suggestion
func (s *Server) Start() error {
       for {
	err := s.runWhileAcquiringDiscoveryGroup(s.start)
	switch{
	case error.Is(...) /* lease was lost*/:
	     continue
	default:
	     return trace.Wrap(err) 
	}
	
      }
}


func (s *Server) start(ctx context.Context) error {
	if s.ec2Watcher != nil {
		go s.handleEC2Discovery()
		go s.reconciler.run(ctx)
		....
```

File: lib/srv/discovery/discovery.go
Comment (tigrato) [inline lib/srv/discovery/discovery.go:1374]:
```suggestion

// acquireDiscoveryGroup tries to acquire a lock if the Service has a DiscoveryGroup
// It will retry using an exponential backoff algorithm.
func (s *Server) runWhileAcquiringDiscoveryGroup(run func(context.Context) error) error {
	if s.DiscoveryGroup == "" {
		s.Log.Warnf("DiscoveryGroup is not set, skipping semaphore lock. It is recommended to set a DiscoveryGroup")
		return trace.Wrap(run(s.ctx))
	}

	s.Log.Debugf("Discovery service is trying to acquire lock for DiscoveryGroup %q", s.DiscoveryGroup)
	lease, err := services.AcquireSemaphoreLock(s.ctx, services.SemaphoreLockConfig{
		Service: s.Config.AccessPoint,
		Expiry:  time.Minute,
		Params: types.AcquireSemaphoreRequest{
			SemaphoreKind: types.SemaphoreKindDiscoveryServiceGroup,
			SemaphoreName: s.DiscoveryGroup,
			MaxLeases:     1,
			Holder:        s.ServerID,
		},
		Clock: s.clock,
	})

	switch {
	case err == nil:
		s.Log.Debugf("Discovery service acquired lock for DiscoveryGroup %q", s.DiscoveryGroup)
		defer lease.Wait()
		defer lease.Stop()
		ctx, cancel := context.WithCancel(s.ctx)
		defer cancel()
		go func() {
			select {
			case <-lease.Done():
				cancel()
			case <-ctx.Done():
				return
			}
		}()
		err := run(ctx)
		switch {
		case errors.Is(err, context.Canceled):
			select {
			case <-s.ctx.Done():
				return trace.Wrap(err)
			default:
				return nil
			}
		}
	case trace.IsAlreadyExists(err) ....:
		s.Log.Debugf("Discovery service is waiting on DiscoveryGroup %q lock", s.DiscoveryGroup)
		return trace.Wrap(err)
	}

	return nil
}

```

File: lib/srv/discovery/discovery.go
Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go:1412]: I'm not sure the service is ready to be started multiple times.
The `runWhileAcquiringDiscoveryGroup` would also need to wait for the service to clean up. Eg watchers, fetchers, installers and reconcilers would need to stop before we can proceed.

File: lib/srv/discovery/discovery.go
Comment (espadolini) [reply lib/srv/discovery/discovery.go:1379]: "Very unlikely" doesn't mean anything; we should set a flag in Stop (while holding the lock) that means "server has been stopped" and we need to check the flag (while holding the lock) as we set `releaseGroupLockFn`, and we should refuse to start if the flag is set (and thus nothing can ever stop the server again).

File: api/types/server.go
Comment (marcoandredinis) [reply api/types/server.go:527]: I'll remove the comment.
EICE nodes already exist and making this a hard requirement would prevent connection to those nodes.

File: lib/srv/discovery/discovery.go
Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go]: Can you please check again?
I've simplified the flow and I'm no longer using the `releaseGroupLockFn`
The AcquireSemaphoreLock receives the Server's Context and should end as soon as we cancel that context.

**Review by espadolini — APPROVED** [review]

File: lib/srv/discovery/discovery.go
Comment (tigrato) [inline lib/srv/discovery/discovery.go:1395]: I don't think you can call s.Stop here.
There are several reasons of it:
- Once you loose a lock you can regain it after a while. This happens several times during auth server restarts where auth is restarted and long-lived connections are droped and the agent loses the lease although the lock is still in its "name". Once auth is back alive the agent can resume the connection. 
- This process is only valid if you have multiple discovery agents running. If you just have one it will cause the discovery service do stop after the first auth disconnection and it won't reconnect
- Stopping the service will cause issues to `/healthz` report as the critical service discovery service won't be running anymore and teleport will report unhealthy. This causes kubernetes to restart pods and load balancers to stop sending traffic.

Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go]: Yeah, you are correct.
I'll remove the semaphore lock logic and rely entirely on server name to ensure we don't duplicate the nodes.

I still think we should invest in preventing duplicate API calls, but I'll leave that to a future PR.

**Comment by marcoandredinis** [comment]
@espadolini Can you please review again? I've removed the semaphore lock logic entirely.
We'll rely on server node name for the purposes of preventing duplicate nodes.

**Review by tigrato — APPROVED** [review]
Can you create an issue to track the service lock? 
It will be important to report the discovery group status so it's worth doing it

**Comment by marcoandredinis** [comment]
Follow up on ensuring a single DiscoveryService is running per DiscoveryGroup https://github.com/gravitational/teleport/issues/40546

File: lib/srv/discovery/discovery.go
Comment (espadolini) [inline lib/srv/discovery/discovery.go:868]: Merge this with the subsequent conditional and do `if existingNode != nil && existingNode.Expiry().After(...) && ...`

**Review by espadolini — APPROVED** [review]
How worried are we that brand new instances will result in an upsert from both auth servers? Is the discovery loop jittered?

File: lib/srv/discovery/discovery.go
Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go]: I actually prefer the current format, but that's fine :+1:

**Comment by marcoandredinis** [comment]
> How worried are we that brand new instances will result in an upsert from both auth servers?

It will happen most of the time for first time that we sync nodes.
After being inserted into cluster, their expiration is jittered and so we'll probably have a single node upsert from one of the servers.

> Is the discovery loop jittered?

No, it's not.
It starts as soon as the service starts, so that should give us a little window where the 2nd server already has the ec2 instances as EICE nodes.
However, after changing a DiscoveryConfig for that DiscoveryGroup, it will re-sync their Fetchers and they will call the APIs at the same™️ time

**Pull request merged**
